docs/design/auditing-containment.md

```diff
@@ -0,0 +1,168 @@
+# Auditing and Node Containment
+
+## Abstract
+
+This design doc outlines how we will implement "containment mode," in which the auditing service monitors a node that has accepted an audit request, but refuses to send the requested data.
+Currently in the codebase, nodes that do this will be marked as having audit failures.
+However, we don't want to mark nodes as offline or immediately mark an audit failure if this specific case arises.
+Instead, we want to "contain" these nodes by retrying the original audit until they either eventually respond and pass the audit, or refuse to respond a certain number of times and ultimately receive an audit failure mark.
+
+## Main Objectives
+1. Nodes won’t be able to “escape” audits for specific data by refusing to send requested data.
+2. The audit service should not audit contained nodes for other data, and should verify their originally requested data.
+3. The overlay service should not select contained nodes.
+4. A ContainmentDB holds metadata required to verify contained nodes' originally requested data.
+
+## Background
+
+The whitepaper section 4.13 talks about containment mode as follows:
+```
+Given a specific storage node, an audit might reveal that it is offline or incorrect. In the case of a node being offline, the audit failure may be due to the address in the node discovery cache being stale, so another, fresh Kademlia lookup will be attempted. If the node still appears to be offline, the Satellite places the node in containment mode. In containment mode, the Satellite will calculate and save the expected response, then continue to try the same audit with that node until the node either responds successfully, actively fails the audit, or is disqualified from being offline too long. Once the node responds successfully, it leaves containment mode. All audit failures will be stored and saved in the reputation system. 
+```
+However, we're departing from the whitepaper because offline nodes will not be moved to containment mode.
+They will just be marked as offline.
+It's only when nodes initially respond to the audit service's dial but then refuse to send the requested erasure share that they are moved to containment mode.
+
+The node will be given a `contained` flag, and if the audit service attempts to audit the node again, it will see this flag and request the same share it had originally requested.
```

Comment:
What happens in the following example:
First audit hits a segment with 80 pieces. For some reason 10 nodes are getting the containment flag.
Second audit hits a segment with 36 pieces. Now we are not allowed to get the pieces from the 10 cointained storage nodes? Why that? What is the advantage of that?

Worst case we end up with only 26 pieces and are not allowed to run this audit even if we could request all 36 pieces.

Reply:
Okay I think you're right, it should be allowed to download the pieces from the contained nodes.